HW2/P2/parallel_vector.pyx

```diff
@@ -74,17 +74,35 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
        int idx, r
        omp_lock_t *locks = get_N_locks(counts.shape[0])
 
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0], num_threads=4):
+               # acquire locks, every element has its own lock
+               # to avoid dead-locking, 
+               # the locking order must be the same between different threads
+               if src[idx] > dest[idx]:
```

Comment:
Usually, you want to do a check again after acquiring the lock, since the condition might have changed between checking the condition and acquiring the lock